### PR TITLE
fix(vector): render GeoPackages without GDAL's WASM thread crash (#393)

### LINKED ISSUE #393: Layers that didn't Work listed on github
I had trouble getting some gpkg and geotif files to you. I think I have them as you suggested in github: https://github.com/muttleytm/muttleytm.github.io

I did try some of the COG TIFs and all but the warped one works. That one really isn't involved in the project I'm interested in at the moment anyway. I continue to not be able to use the gpkg layers. As before a few work, but many don't. That was trying to add them to the web version of Geo Libre.

I see Geo Libre works with other layer formats. I may also try exporting them in other formats and see how that goes. I have noticed that with the gpkg layers that did work I don't see any label on the layer or how to set one. Maybe I missed a setting when I exported as a gpkg or something. I will also look into that.

### PR DESCRIPTION
**Pull request by giswqs**
## Problem

Closes #393. Many valid GeoPackages fail to render in the web app (a few small ones happen to work). Loading them throws:

```
Invalid Error: thread constructor failed: Resource temporarily unavailable
```

### Root cause

DuckDB-WASM's `ST_Read` (GDAL) reads a GeoPackage layer by filling Arrow result batches on a **background thread** once the layer crosses a small data threshold. The single-threaded DuckDB-WASM `eh` bundle the app loads has no pthread support, so `std::thread` aborts with EAGAIN. I reproduced this with the exact deployed `@duckdb/duckdb-wasm@1.33.1-dev45.0` + spatial extension: `DESCRIBE` and `ST_Read_Meta` succeed, but any feature read (`LIMIT 1`, `count(*)`, `SELECT *`) crashes once the result is more than a few features.

The existing `gpkg_ogr_contents` repair (#258/#376) fixes the related feature-**count** path, but not this feature-**read** path, and these files already carry valid counts. No DuckDB-reachable GDAL config (open options, `max_batch_size`, `keep_wkb`, `threads`) disables the threading.

## Fix

Read GeoPackages directly from their SQLite tables with **sql.js** (already bundled for the count repair) instead of going through GDAL:

- A new `decodeWkb` (the exact inverse of the existing `encodeWkb`) decodes the standard WKB inside each GeoPackage geometry blob to GeoJSON. Handles mixed byte order, ISO Z/M dimensionality, and EWKB flags; throws clearly on curved types GeoJSON cannot represent.
- `gpkg-reader.ts` selects the first feature layer, strips the `GP` blob header, decodes each geometry, and excludes the integer primary key from properties.
- Non-WGS84 layers are reprojected to WGS84 through the **existing** DuckDB `ST_Transform` path, whose GeoJSON reader is not affected by the threading bug (verified at 300 features).
- The large-dataset guard is preserved (counts rows first).

`ST_Read` is still used for every other vector format and for the GeoParquet conversion path; a file mislabelled `.gpkg` that is not actually SQLite falls through to `ST_Read`.

## Verification

- **Clean-room repro** of the crash using the deployed DuckDB-WASM + spatial versions.
- **Byte-identical geometry**: the new reader's output matches `ogr2ogr` ground truth with 0 geometry and 0 property mismatches across all four files in the issue (17 LineStrings, 182 + 19 Polygons, 275 LineStrings), plus MultiPolygon, 3D (Z), and EPSG:3857 cases.
- **End to end in the real app** (production build + drag-and-drop): all four GeoPackages render their geometry on the map with populated attribute tables and no console errors.
- 959 frontend tests pass (20 new for the decoder and reader); typecheck and scoped `pre-commit` (eslint + npm build) pass.

## Notes

The warped COG from the issue is unrelated to this rendering path and the reporter said it is out of scope for now, so it is not addressed here.

## Summary by CodeRabbit

* **New Features**
  * Added dedicated GeoPackage (.gpkg) loading with automatic layer detection, direct table reads, WKB-to-GeoJSON decoding, property cleanup, and EPSG resolution with optional reprojection.
  * Added WKB decoding for common geometry types with correct endianness/dimensionality handling and explicit errors for unsupported curved types.
  * Added guards for large GeoPackage datasets during reads.
* **Tests**
  * Added unit tests for WKB decoding and GeoPackage reading, including envelope/header edge cases and malformed blobs.
* **Chores**
  * Updated the raster map dependency to the latest patch version.

---

## Also in this PR: the warped COG from #393

Issue #393 also reported one COG (`warped1868mapv1testexport1COG.tif`) that would not render while the others did. Root cause: its `GTModelTypeGeoKey` is `32767` ("user-defined") even though the projection is fully described by the geo-keys (it is an ArcGIS export of the ESRI `WGS_1984_Web_Mercator` auxiliary-sphere Mercator, which carries no EPSG code). The CRS parser threw `Unsupported GeoTIFF model type: 32767`, so the layer never drew.

That fix lives in the `maplibre-gl-raster` package (opengeos/maplibre-gl-raster#10, released as `0.2.2`): `loadGeoTIFF` now normalises a user-defined model type to "projected" when a projection method is present, so the CRS is built from the keys already in the file. The fix is projection-agnostic.

This PR bumps `maplibre-gl-raster` to `^0.2.2` so GeoLibre picks it up. Verified end to end: the previously invisible warped COG now renders at the correct location.

### CODE REVIEW

File: apps/geolibre-desktop/src/lib/geometry-wkb.ts [hunk 1]
```diff
@@ -131,6 +131,140 @@ export function encodeWkb(geometry: Geometry): Uint8Array {
...
+    // any standards-conformant GeoPackage geometry blob decodes.
+    const hasEwkbZ = (rawType & 0x80000000) !== 0;
+    const hasEwkbM = (rawType & 0x40000000) !== 0;
+    const hasSrid = (rawType & 0x20000000) !== 0;
+    const baseType = rawType & 0xffff;
+    const isoGroup = Math.floor((baseType % 4000) / 1000);
+    const code = baseType % 1000;
+    const hasZ = hasEwkbZ || isoGroup === 1 || isoGroup === 3;
+    const hasM = hasEwkbM || isoGroup === 2 || isoGroup === 3;
+    const dimensions = 2 + (hasZ ? 1 : 0) + (hasM ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/geometry-wkb.ts]: Done in 19a4886 — removed the unused `dimensions` variable.

File: apps/geolibre-desktop/src/lib/geometry-wkb.ts [hunk 2]
```diff
@@ -131,6 +131,140 @@ export function encodeWkb(geometry: Geometry): Uint8Array {
...
+          ),
+        };
+      }
+      case 7:
+        return { type: "GeometryCollection", geometries: readChildren() };
+      default:
+        throw new Error(
+          `Unsupported WKB geometry type ${code}${
+            WKB_TYPE_NAME[code] ? "" : " (curved geometries are not supported)"
+          }.`,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/geometry-wkb.ts:253]: Done in 19a4886 — the curved-geometry note is now appended only for type codes 8-12; removed the always-false `WKB_TYPE_NAME` ternary (and the now-unused map).

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+ * unchanged so non-conformant producers still read.
+ */
+export function stripGeoPackageHeader(blob: Uint8Array): Uint8Array {
+  // 'G','P' magic identifies a GeoPackage geometry blob; otherwise assume the
+  // value is already standalone WKB (byte-order byte 0x00 or 0x01).
+  if (blob.length < 2 || blob[0] !== 0x47 || blob[1] !== 0x50) return blob;
+  const flags = blob[3];
+  const envelopeIndicator = (flags >> 1) & 0x07;
+  // Envelope byte sizes by indicator: 0=none, 1=XY, 2=XYZ, 3=XYM, 4=XYZM.
+  const envelopeBytes = [0, 32, 48, 48, 64, 0, 0, 0][envelopeIndicator];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts]: Done in 19a4886 — `stripGeoPackageHeader` now throws on a reserved envelope indicator (5-7) instead of mapping it to 0 and decoding from the wrong offset.

File: apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts [hunk 4]
```diff
@@ -312,10 +317,55 @@ async function countFeatures(
...
+    const counted = await countGeoPackageFeatures(file.data);
+    if (counted) {
+      await confirmLargeDataset(
+        { name: file.name, featureCount: counted.featureCount },
+        options.onLargeDataset,
+      );
+    }
+  }
+
+  const { featureCollection, epsgCode } = await loadGeoPackageVectorFile(
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts]: Done in 19a4886 — `loadGeoPackageVectorFile` now takes the large-dataset guard and counts on the same DB open, so a counted load no longer opens/parses the file twice.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  idColumn: string | null;
+}
+
+export interface GeoPackageReadResult {
+  featureCollection: FeatureCollection<Geometry | null>;
+  /**
+   * The EPSG code the geometries are stored in, or null when they are already
+   * WGS84 lon/lat (or the CRS is undefined). The caller reprojects when set.
+   */
+  epsgCode: number | null;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:42]: Done in 19a4886 — `quoteIdentifier` and `tableExists` are now exported from `gpkg-ogr-contents.ts` and imported here instead of duplicated.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+export function stripGeoPackageHeader(blob: Uint8Array): Uint8Array {
+  // 'G','P' magic identifies a GeoPackage geometry blob; otherwise assume the
+  // value is already standalone WKB (byte-order byte 0x00 or 0x01).
+  if (blob.length < 2 || blob[0] !== 0x47 || blob[1] !== 0x50) return blob;
+  const flags = blob[3];
+  const envelopeIndicator = (flags >> 1) & 0x07;
+  // Envelope byte sizes by indicator: 0=none, 1=XY, 2=XYZ, 3=XYM, 4=XYZM.
+  const envelopeBytes = [0, 32, 48, 48, 64, 0, 0, 0][envelopeIndicator];
+  const headerLength = 8 + envelopeBytes;
+  return blob.subarray(headerLength);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:78]: Done in 19a4886 — `stripGeoPackageHeader` now validates the header: it throws on a truncated header, a truncated envelope, or a reserved envelope indicator (5-7), so a malformed blob is an explicit error rather than a silent wrong/null geometry.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      throw new Error("No vector feature layer found in this GeoPackage.");
+    }
+    const epsgCode = resolveEpsgCode(db, layer.srsId);
+
+    const result = db.exec(`SELECT * FROM ${quoteIdentifier(layer.table)}`);
+    const features: Feature<Geometry | null>[] = [];
+    if (result.length > 0) {
+      const columns = result[0].columns;
+      const geometryIndex = columns.indexOf(layer.geometryColumn);
+      const idIndex = layer.idColumn ? columns.indexOf(layer.idColumn) : -1;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts]: Done in 19a4886 — `readLayerFeatures` now throws when the declared geometry column is absent from SELECT * instead of emitting every feature with a null geometry.

File: tests/gpkg-reader.test.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    blob[3] = 0b0000_0010; // envelope indicator 1 (XY) in bits 1-3
+    blob.set(wkb, 8 + 32);
+    assert.deepEqual([...stripGeoPackageHeader(blob)], [...wkb]);
+  });
+
+  it("returns bare WKB unchanged", () => {
+    const wkb = encodeWkb({ type: "Point", coordinates: [1, 2] });
+    assert.deepEqual([...stripGeoPackageHeader(wkb)], [...wkb]);
+  });
+});
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/gpkg-reader.test.ts:137]: Done in 19a4886 — added regression tests for malformed geometry headers (truncated header/envelope, reserved envelope indicator) and a missing declared geometry column.

File: apps/geolibre-desktop/src/lib/geometry-wkb.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/geometry-wkb.ts]: Already done in 19a4886 — the unused `dimensions` variable was removed. (This thread is outdated against the pre-fix commit.)

File: apps/geolibre-desktop/src/lib/geometry-wkb.ts [hunk 9]
```diff
@@ -131,6 +131,140 @@ export function encodeWkb(geometry: Geometry): Uint8Array {
...
+            (p) => (p as { coordinates: Position[][] }).coordinates,
+          ),
+        };
+      }
+      case 7:
+        return { type: "GeometryCollection", geometries: readChildren() };
+      default:
+        throw new Error(
+          `Unsupported WKB geometry type ${code}${
+            WKB_TYPE_NAME[code] ? "" : " (curved geometries are not supported)"
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/geometry-wkb.ts]: Already done in 19a4886 — the curved-geometry note is now appended only for codes 8-12 and the always-false `WKB_TYPE_NAME` map was removed. (Outdated against the pre-fix commit.)

File: apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts [hunk 10]
```diff
@@ -312,10 +317,55 @@ async function countFeatures(
...
+    file.data,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts]: Already done in 19a4886 — `loadGeoPackageVectorFile` now counts on the same DB open as the read, so the file is no longer parsed twice. (Outdated against the pre-fix commit.)

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+interface GeoPackageLayer {
+  table: string;
+  geometryColumn: string;
+  srsId: number | null;
+  /** The INTEGER PRIMARY KEY column, excluded from feature properties. */
+  idColumn: string | null;
+}
+
+export interface GeoPackageReadResult {
+  featureCollection: FeatureCollection<Geometry | null>;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:37]: Already done in 19a4886 — `quoteIdentifier` and `tableExists` are exported from `gpkg-ogr-contents.ts` and imported here; the local copies were removed.

File: apps/geolibre-desktop/package.json [hunk 12]
```diff
@@ -67,7 +67,7 @@
     "maplibre-gl-nasa-earthdata": "^0.1.2",
     "maplibre-gl-national-map": "^0.1.1",
     "maplibre-gl-planetary-computer": "^0.3.0",
-    "maplibre-gl-raster": "^0.2.1",
+    "maplibre-gl-raster": "^0.2.2",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/package.json]: `maplibre-gl-raster@0.2.2` is published — it was released minutes before this push (npm `latest` is 0.2.2, and pre-commit's `npm install` + `npm run build` resolved it from the registry and built successfully). The web-query snapshot in this comment is stale (it still reports 0.2.0). The integration uses only the public `addRaster` API, not the private internals listed. No change needed.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+     FROM gpkg_geometry_columns g
+     JOIN gpkg_contents c ON c.table_name = g.table_name
+     WHERE lower(c.data_type) = 'features'
+     ORDER BY c.rowid
+     LIMIT 1`,
+  );
+  const row = result[0]?.values[0];
+  if (!row) return null;
+  const table = String(row[0]);
+  const geometryColumn = String(row[1]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:107]: Done in 7871a90 — `selectLayer` now guards `gpkg_contents` existence before the JOIN, returning null for a malformed file instead of throwing an opaque sql.js error.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+          geometry = wkb.length > 0 ? decodeWkb(wkb) : null;
+        }
+        features.push({ type: "Feature", geometry, properties });
+      }
+    }
+
+    return {
+      featureCollection: { type: "FeatureCollection", features },
+      epsgCode,
+    };
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts]: Done in 7871a90 — per-feature geometry decode is now wrapped in try/catch: one unreadable blob keeps its feature with a null geometry and logs a warning, instead of aborting the whole layer. (stripGeoPackageHeader still throws on a malformed header; the per-feature catch turns that into a warned skip.)

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+  return blob.subarray(headerLength);
+}
+
+/**
+ * Pick the layer to read: the first `features` row in `gpkg_contents` that has a
+ * registered geometry column. Mirrors GDAL's "first layer" default (no `layer=`
+ * argument), which is what the previous `ST_Read` path used.
+ */
+function selectLayer(db: Database): GeoPackageLayer | null {
+  // Both tables are referenced by the JOIN below. gpkg_contents is mandatory in
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:87]: Already done in 7871a90 — `selectLayer` now guards `gpkg_contents` existence before the JOIN (this round reviewed the prior commit).

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        const value = row[i];
+        // sql.js returns BLOB columns as Uint8Array; binary attributes are not
+        // JSON-serialisable, so drop them (matching the ST_Read path).
+        if (value instanceof Uint8Array) continue;
+        properties[columns[i]] = value;
+      }
+
+      const rawGeometry = row[geometryIndex];
+      let geometry: Geometry | null = null;
+      if (rawGeometry instanceof Uint8Array && rawGeometry.length > 0) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:193]: Already done in 7871a90 — the per-feature geometry decode is wrapped in try/catch: one unreadable blob keeps its feature with a null geometry and warns, matching the ST_Read behaviour you describe (this round reviewed the prior commit).

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+  // srs_id 0 = undefined geographic, -1 = undefined cartesian (GeoPackage spec).
+  if (srsId == null || srsId === 0 || srsId === -1 || WGS84_EPSG_CODES.has(srsId)) {
+    return null;
+  }
+  if (!tableExists(db, "gpkg_spatial_ref_sys")) return null;
+  const row = db.exec(
+    `SELECT organization, organization_coordsys_id
+     FROM gpkg_spatial_ref_sys WHERE srs_id = :id`,
+    { ":id": srsId },
+  )[0]?.values[0];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:139]: Done in f36b2ba — resolveEpsgCode now rejects a non-finite code (`!Number.isFinite(code)`), so a non-numeric organization_coordsys_id resolves to null (no reprojection) instead of tagging the collection EPSG:NaN. Added a regression test.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+    featureCollection: readLayerFeatures(db, layer),
+    epsgCode: resolveEpsgCode(db, layer.srsId),
+  };
+}
+
+/**
+ * Synchronous core of {@link loadGeoPackageVectorFile}: read every feature of
+ * the first layer into a GeoJSON FeatureCollection. Separated so it can be
+ * unit-tested with an already-initialised sql.js factory.
+ */
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:231]: Done in f36b2ba — renamed to `isLikelyGeoPackage` and expanded the docstring to make clear it only checks the SQLite magic, not the GeoPackage schema.

File: tests/gpkg-reader.test.ts [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+
+/** Wrap WKB in a minimal little-endian GeoPackage geometry blob (no envelope). */
+function geoPackageBlob(geometry: Geometry, srsId = 4326): Uint8Array {
+  const wkb = encodeWkb(geometry);
+  const blob = new Uint8Array(8 + wkb.length);
+  const view = new DataView(blob.buffer);
+  blob[0] = 0x47; // 'G'
+  blob[1] = 0x50; // 'P'
+  blob[2] = 0x00; // version
+  blob[3] = 0x01; // flags: little-endian header, no envelope
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/gpkg-reader.test.ts:30]: I think this one is the other way round per the GeoPackage spec (OGC 12-128r18, GeometryHeaderFlags): **bit 0** of the flags byte is the byte order of the header's srs_id/envelope fields (0 = big-endian, 1 = little-endian), and the **GeoPackageBinary type** (standard vs extended) is **bit 5** (0x20). So `0x01` correctly marks a little-endian, standard, no-envelope blob, and the comment is accurate. `0x00` would declare a big-endian header, which would then disagree with the little-endian srs_id this helper writes. Leaving the comment as-is, but happy to adjust if I'm misreading the spec.

File: tests/geometry-wkb-decode.test.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    writer.setFloat64(21, 100, true); // Z (kept)
+    writer.setFloat64(29, 7, true); // M (dropped)
+    const decoded = decodeWkb(new Uint8Array(writer.buffer));
+    assert.deepEqual(decoded, { type: "Point", coordinates: [-85.6, 42.9, 100] });
+  });
+
+  it("throws on unsupported curved geometry types", () => {
+    // Type code 8 = CircularString, which GeoJSON cannot represent.
+    const bytes = new Uint8Array([0x01, 0x08, 0x00, 0x00, 0x00]);
+    assert.throws(() => decodeWkb(bytes), /Unsupported WKB geometry type 8/);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/geometry-wkb-decode.test.ts:93]: Done in 8bc51ba — added a test covering the EWKB SRID-prefix skip branch in decodeWkb.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+function selectLayer(db: Database): GeoPackageLayer | null {
+  // Both tables are referenced by the JOIN below. gpkg_contents is mandatory in
+  // the spec, but guard it too so a malformed file returns null (no layer) here
+  // rather than throwing an opaque sql.js error from the JOIN.
+  if (!tableExists(db, "gpkg_geometry_columns")) return null;
+  if (!tableExists(db, "gpkg_contents")) return null;
+  const result = db.exec(
+    `SELECT g.table_name, g.column_name, g.srs_id
+     FROM gpkg_geometry_columns g
+     JOIN gpkg_contents c ON c.table_name = g.table_name
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts]: Done in 8bc51ba — the JOIN now uses `COLLATE NOCASE`, so a table name spelled with different casing in gpkg_contents vs gpkg_geometry_columns still resolves a layer.

File: apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts [hunk 22]
```diff
@@ -312,10 +313,56 @@ async function countFeatures(
...
+          options.onLargeDataset,
+        )
+    : undefined;
+
+  const { featureCollection, epsgCode } = await loadGeoPackageVectorFile(
+    file.data,
+    onBeforeRead,
+  );
+  if (epsgCode == null) {
+    return featureCollection as FeatureCollection;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/duckdb-vector-loader.ts:342]: This mirrors the existing ST_Read path, which also returns `toFeatureCollection(...) as FeatureCollection`. Null-geometry features are valid throughout the app (the store and map ignore null geometries) and `loadDuckDbVectorFile`'s public return type is `FeatureCollection`, so widening only the GeoPackage path would be inconsistent. Keeping the cast for parity with the existing reader.

File: apps/geolibre-desktop/src/lib/gpkg-reader.ts [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+  const table = String(row[0]);
+  const geometryColumn = String(row[1]);
+  const srsId = row[2] == null ? null : Number(row[2]);
+
+  let idColumn: string | null = null;
+  for (const info of db.exec(
+    `PRAGMA table_info(${quoteIdentifier(table)})`,
+  )[0]?.values ?? []) {
+    // table_info columns: cid, name, type, notnull, dflt_value, pk.
+    if (info[5] === 1) idColumn = String(info[1]);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/lib/gpkg-reader.ts:115]: GeoPackage feature tables are required by the spec to have a single-column integer primary key, so a composite PK is out of spec. The first PK column (the fid/rowid alias) is excluded; any extra columns from a malformed composite key fall through as attributes, which is a reasonable degradation for a non-conformant file rather than something worth special-casing.
